### PR TITLE
Range casts not to throw

### PR DESCRIPTION
**Pull request by vasslitvinov**
This PR implements the changes proposed in https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1678271195 and https://github.com/chapel-lang/chapel/issues/19269#issuecomment-1685042219

Specifically in the range cast:
* cast the bounds without any runtime checks; ex. `(a..b): range(int(8))` is simply `(a: int(8)) .. (b: int(8))`
  - the range cast still throws when the cast on the individual bounds throws, for example when (unstably) casting from a range of ints to a range of enums
* cast the stride with safeCast-style checks:
  - if the original stride does not fit in the new idxType or strideKind, then halt
  - these checks are disabled upon --fast or, respectively, upon --no-cast-checks and --no-bounds-checks
* introduce `tryCast` that always checks everything at runtime and `throws` when a check is violated
  - `tryCast` does exactly what the cast operator did prior to this PR

The domain cast:
* continues to invoke the range cast in each dimension
* introduce `tryCast` that invokes the range `tryCast` in each dimension

Testing: passes standard and gasnet paratests.

### CODE REVIEW

File: modules/internal/ChapelRange.chpl [hunk 1]
```diff
@@ -1722,6 +1705,76 @@ operator :(r: range(?), type t: range(?)) throws where !chpl_castIsSafe(r, t) {
   return tmp;
 }
 
+/* Cast a range to a new range type.  The overload below throws when
+   the original bounds and/or stride do not fit in the new type or 'strides'.
+   TODO: should we allow 't' to be generic?
+ */
+@chpldoc.nodoc
+operator :(r: range(?), type t: range(?)) where chpl_castIsSafe(r, t) {
+  // note: the implementation is duplicated with the throwing overload
```
Comment (jabraham17) [inline modules/internal/ChapelRange.chpl:1714]: If this is the same as tryCast, why not call `tryCast`?

File: modules/internal/ChapelRange.chpl [hunk 2]
```diff
@@ -1722,6 +1705,76 @@ operator :(r: range(?), type t: range(?)) throws where !chpl_castIsSafe(r, t) {
   return tmp;
 }
 
+/* Cast a range to a new range type.  The overload below throws when
+   the original bounds and/or stride do not fit in the new type or 'strides'.
+   TODO: should we allow 't' to be generic?
+ */
```
Comment (jabraham17) [inline modules/internal/ChapelRange.chpl:1711]: This has a doc comment but is marked nodoc?

File: test/types/range/enum/castEnumRange4a.good [hunk 3]
```diff
@@ -1,6 +1,5 @@
 castEnumRange4a.chpl:3: warning: Casts between ranges involving 'enum' indices are currently unstable (see issue #22406); consider performing the conversion manually
 castEnumRange4a.chpl:5: warning: Casts between ranges involving 'enum' indices are currently unstable (see issue #22406); consider performing the conversion manually
-castEnumRange4a.chpl:7: warning: Casts between ranges involving 'enum' indices are currently unstable (see issue #22406); consider performing the conversion manually
```
Comment (jabraham17) [inline test/types/range/enum/castEnumRange4a.good:3]: why does this no longer warn? should it?

**Review by jabraham17 — APPROVED** [review]
Looks good!

File: modules/internal/ChapelRange.chpl (hunk 1 above)
Comment (vasslitvinov) [reply modules/internal/ChapelRange.chpl:1769]: It was `operator :` prior to this PR that behaved the same as the new `tryCast`. With this PR, `operator :` and `tryCast` behave differently.

File: modules/internal/ChapelRange.chpl (hunk 2 above)
Comment (vasslitvinov) [reply modules/internal/ChapelRange.chpl:1766]: It was this way a long time ago. My understanding is that this comment is aimed at the developers of ChapelRange, not users of ranges.

File: test/types/range/enum/castEnumRange4a.good (hunk 3 above)
Comment (vasslitvinov) [reply test/types/range/enum/castEnumRange4a.good:3]: This is due to a bug in how our compiler generates warnings, which got triggered by factoring out the warnings into a helper function, namely `checkEnumIdx`.

Until that bug is fixed, we have the tradeoff between 4x code duplication vs. potentially-missing unstable warnings.

Meanwhile I added a .future test, `test/types/range/enum/castEnumRange4b`

**Pull request merged**